lib/core/src/Cardano/Wallet/Primitive/Types/TokenPolicy.hs

```diff
@@ -5,10 +5,10 @@
 module Cardano.Wallet.Primitive.Types.TokenPolicy
     (
       -- * Token Policies
-      TokenPolicyId
+      TokenPolicyId (..)
```

Comment:
@Anviking

Just a thought.

Rather than exporting the default data constructors for `TokenPolicyId` and `TokenName`, perhaps we could define and export smart constructors instead?

It would be my prefererence to not export these default constructors, as it opens the door to people accidentally creating invalid token policy IDs (which should have a certain length), and this could cause subtle regressions in other places.

In the case of `TokenName`, there is actually already a `FromText` instance. Right now, this never fails, but if token names were to require validation, then calling `fromText` would be safer than using the default constructor.

Reply:
> It would be my preference to not export these default constructors, as it opens the door to people accidentally 

👍 Good catch, agree.

Though, I wonder what the best solution is 🤔 

> were to require validation, then calling fromText would be safer than using the default constructor.

If fromText fails in the node-to-wallet type-conversion, then the wallet and node have different validation rules, and surely the node's must be the correct ones. Do we actually want to validate in the conversion?

Reply:
I exported those constructors, because it's impractical not to have them. PolicyId and AssetName values coming from the node are already validated.

Reply:
> I exported those constructors, because it's impractical not to have them. PolicyId and AssetName values coming from the node are already validated.

Well, one alternative is to create functions `unsafePolicyId` and `unsafePolicyName`. 

Advantages:

1. Callers of these functions are explicitly reminded that they're using unsafe functions. They onus is on them to check that the input is valid.

2. When people reviewing code see a new call to `unsafeX`, they can ask the question: "Has this input been validated?"

3. We can easily track usages of `unsafe` functions with a simple call to `grep` (or `ag`). This is a bit harder to do with the default data constructor.

Reply:
The type constructor itself could be called UnsafePolicyId if you like.

Reply:
> The type constructor itself could be called UnsafePolicyId if you like.

That's not a bad alternative (it passes the `grep` test), but it alters the output from `show`. So instead of:

```
PolicyId "deaddead....
```

We'd have:
```
UnsafePolicyId "deaddead....
```

It's a minor issue of course, but in a structure with lots of policy IDs, this increased length does tend to add up.

Reply:
@jonathanknowles I think we do not care much about that last bit :blush: ... if that really bothers you: https://hackage.haskell.org/package/quiet-0.2/docs/Quiet.html

Reply:
> @jonathanknowles I think we do not care much about that last bit blush ...

Fair enough.

> if that really bothers you: https://hackage.haskell.org/package/quiet-0.2/docs/Quiet.html

We actually are _already_ using `Quiet` here:

```
newtype TokenPolicyId = TokenPolicyId
    { unTokenPolicyId :: Hash "TokenPolicy" }
    deriving stock (Eq, Ord, Generic)
    deriving (Read, Show) via (Quiet TokenPolicyId)
```

Reply:
I would have been slightly in favour of `unsafePolicyId` over `UnsafePolicyId` for the sake of:
- Less confusing `Show` output
- No unintended **de**-construction outside the module

Reply:
@Anviking we also need **de**-construction in Compatibility.hs.

The situation is a bit ridiculous with the multiple layers of wrapping around a 32 byte bytestring (maximum length not enforced by our code by the way). And Text is probably not the correct inner type anyway!

Reply:
Ok, fair, perhaps.